### PR TITLE
969 - TYDE CoS #6 - Create circle view

### LINKED ISSUE #969: TYDE CoS #6 - Create circle view

### PR DESCRIPTION
**Pull request by narendrasss**
## Changes
* Implement the circle view

## Approach
* Data for the circle view is encoded as an array of arrays in the `cos` object. Each array contains a list of connection IDs that correspond to the connections that live in that circle. The array at index 0 represents the innermost circle, the array at index 1 represents the second innermost, and so on.
* Each circle is an `li` element with a fixed width and height calculated solely from the index of the array that it represents. For example, the array at index 0 represents the innermost circle, so it has the smallest radius. **This should be changed to depend on other factors, including the number of connections in the array and the current viewport size**.
* Each connection within a circle is positioned using its polar coordinates. These coordinates are converted to cartesian coordinates and applied using transform.

## Screenshot
https://user-images.githubusercontent.com/31267630/110851959-9a2f5380-8266-11eb-9f2a-e2d0cefe10d3.mp4

## Issue Linkage
Closes #969

## PR Dependency
Depends on: #952 

## Notes

- There's a bomb button that I added in for testing purposes. When pressed, the bomb will wipe the CoS data, letting you start over from scratch. This will be removed once testing is done.

### CODE REVIEW

**Comment by narendrasss** [comment]
@wynnset This is ready for testing, but it looks like the CI is broken again for some reason 😢

**Review by shirleywbi — COMMENTED** [review]
@narendrasss Some questions/comments:
1. When I opened this, there was a bomb in the middle. When I added my first connection, the connection appeared in the center. Is this expected? 
image
2. From a ux perspective, do we need to add instructions on how to use the circle view? It took me a solid few minutes to figure out it was drag and drop and only when I realized it was how close you are with a connection.
3. Should we add a warning for when the bomb is clicked so you don't accidentally delete all your connections and communities?
4. Is the bomb supposed to be on both the circle view and the tab-ish view?
5. Should we increase the z-index for the names (left) and add a background color (right)?
image

**Comment by narendrasss** [comment]
@shirleywbi Oops! Ignore the bomb - it's only for testing purposes since there's currently no way to delete communities once you've added them.

1. When I opened this, there was a bomb in the middle. When I added my first connection, the connection appeared in the center. Is this expected?

The face in the center isn't a connection, it's meant to be the current user's profile/avatar. Right now we don't have a profile implemented so I'm using that emoji as a placeholder.

2. From a ux perspective, do we need to add instructions on how to use the circle view? It took me a solid few minutes to figure out it was drag and drop and only when I realized it was how close you are with a connection.

I think this is part of task 7 with the on-boarding task.

3. Should we increase the z-index for the names (left) and add a background color (right)?

Good catch!

**Review by shirleywbi — COMMENTED** [review]
@narendrasss Came across some strange behaviour while testing

1. Emoji ended up in a different circle than where I dragged it to.
image
image

2. Can we adjust the circle to fit all of the circle in the box?
image

3. On a more empty circle, I'm finding that moving the emojis around and having the emojis disappear depending on which ring its in is a little 

4. Not sure if we want to fix this but I'm having trouble with the lack of feedback on an emoji successfully getting added when it appears under the connections tab. For instance, if you already have one connection in a circle, the next position it adds to is the bottom which is covered, but the connections tab is still open so you don't actually get any feedback whatsoever. Would it be better to have the connections tab reappear after the dragged emojis appears on the circle?
image

5. When the connections tab is up and you're dragging the emojis from the circle to elsewhere in the circle, the connections tab stays up. Should this disappear like how it does when you're dragging from the connections tab to the circle?

File: endpoints/endpoints.circle-of-support.php [hunk 1]
```diff
@@ -73,4 +79,30 @@ public static function removeConnectionFromCommunity($request)
...
+    public static function addConnectionToCircle($request)
+    {
+        $circleIndex = $request['circleIndex'];
+        $cos = new CircleOfSupport();
+        $circle = $cos->addConnectionToCircle(
+            $circleIndex,
+            json_decode($request->get_body())->id
+        );
+        $cos->save();
+        return $circle;
```
Comment (shirleywbi) [inline endpoints/endpoints.circle-of-support.php:92]: Missing try-catch?
```suggest
        try {
            $circleIndex = $request['circleIndex'];
            $connectionId = $request['connectionId'];
            $cos = new CircleOfSupport();
            $circle = $cos->addConnectionToCircle(
                    $circleIndex,
                    json_decode($request->get_body())->id
             );
            $cos->save();
        } catch (TapestryError $e) {
            return new WP_Error($e->getCode(), $e->getMessage(), $e->getStatus());
        }

File: templates/vue/src/components/tyde/activities/CircleOfSupport/CircleView/CircleToggle.vue [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      :class="['circle', { active: state === states.All || state === states.Two }]"
+    />
+    <span
+      :class="['circle', { active: state === states.All || state === states.Three }]"
+    />
+  </button>
+</template>
+
+<script>
+const States = {
```
[the remark this answers is not in the document]
Comment (narendrasss) [reply templates/vue/src/components/tyde/activities/CircleOfSupport/CircleView/CircleToggle.vue:16]: Sure, I'll add comments.

**Comment by narendrasss** [comment]
Addressed most of the UX issues. @shirleywbi Let me know if this is better:

https://user-images.githubusercontent.com/31267630/112365190-97842380-8c94-11eb-8508-ad24e93c2452.mp4

**Review by shirleywbi — APPROVED** [review]
@narendrasss Thanks for the changes. Looks good!

**Comment by wynnset** [comment]
@narendrasss please fix the merge conflicts on this.

**Comment by wynnset** [comment]
@narendrasss the circles are not showing up for me in the circle view. Can you double-check please?

**Comment by narendrasss** [comment]
@wynnset Did you happen to use this same CoS for testing the previous branches? There was an issue with the initialization of previous CoS that would lead to the circles missing. Can you try testing this on a new account to see if it still happens?

**Comment by wynnset** [comment]
As discussed, let's:
- [ ] Highlight the circle (shade it grey?) when selecting a particular circle (circle view)
- [ ] Fix the community edit box to show beside the emoji for the community, not in the top left corner like it is now (community view)

**Comment by narendrasss** [comment]
@wynnset made the changes. Let me know if there's anything else.

**Review by wynnset — APPROVED** [review]

**Pull request merged**